src/circuit/operator/differentiatior.rs

```diff
@@ -0,0 +1,57 @@
+// Differentiator operator: direct implementation, without an explicit z
+
+use crate::algebra::GroupValue;
+use crate::circuit::operator_traits::{Operator, UnaryOperator};
+use std::borrow::Cow;
+
+pub struct Differentiator<V>
+where
+    V: GroupValue,
+{
+    previous: V,
+}
+
+impl<V> Differentiator<V>
+where
+    V: GroupValue,
+{
+    #[allow(dead_code)]
+    pub fn new() -> Self {
+        Self {
+            previous: V::zero(),
+        }
+    }
+}
+
+impl<V> Default for Differentiator<V>
+where
+    V: GroupValue,
+{
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<V> Operator for Differentiator<V>
+where
+    V: GroupValue,
+{
+    fn name(&self) -> Cow<'static, str> {
+        Cow::from("Differentiator")
+    }
+    fn clock_start(&mut self) {}
+    fn clock_end(&mut self) {
+        self.previous = V::zero();
+    }
+}
+
+impl<V> UnaryOperator<V, V> for Differentiator<V>
+where
+    V: GroupValue,
+{
+    fn eval(&mut self, i: &V) -> V {
+        let result = i.add_by_ref(&self.previous.neg_by_ref());
+        self.previous = i.clone();
+        result
+    }
```

Comment:
I think we can do better using `neg` instead of `neg_by_ref`:

```
    fn eval(&mut self, i: &V) -> V {
        let mut old = i.clone();
        swap(&mut old, &mut self previous);
        self.previous.add_by_ref(&self.old.neg());
        self.previous = i.clone();
        result
    }
```

Reply:
We should also write an `eval_owned` that avoids cloning `i`.